src/hpr_routing.erl

```diff
@@ -0,0 +1,112 @@
+-define(UNCONFIRMED_UP, 2#010).
+-define(CONFIRMED_UP, 2#100).
+
+-spec init() -> ok.
+init() ->
+    HotspotRateLimit = application:get_env(router, hotspot_rate_limit, ?DEFAULT_HOTSPOT_THROTTLE),
+    ok = throttle:setup(?HOTSPOT_THROTTLE, HotspotRateLimit, per_second),
+    ok.
+
+-spec handle_packet(Packet :: hpr_packet_up:packet(), HandlerPid :: pid()) -> ok | {error, any()}.
+handle_packet(Packet, HandlerPid) ->
+    Checks = [
+        {fun hpr_packet_up:verify/1, bad_signature},
+        {fun throttle_check/1, hotspot_limit_exceeded}
+    ],
+    case execute_checks(Packet, Checks) of
+        {error, _} = Error ->
+            HandlerPid ! Error,
+            Error;
+        ok ->
+            case packet_type(Packet) of
+                undefined ->
+                    Error = {error, invalid_packet_type},
+                    HandlerPid ! Error,
+                    Error;
+                {join_req, AppEUI, DevEUI} ->
+                    Routes = hpr_routing_config_worker:lookup_eui(AppEUI, DevEUI),
+                    ok = deliver_packet(Packet, HandlerPid, Routes);
+                {uplink, DevAddr} ->
+                    Routes = hpr_routing_config_worker:lookup_devaddr(DevAddr),
+                    ok = deliver_packet(Packet, HandlerPid, Routes)
+            end
+    end.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+-spec deliver_packet(
+    Packet :: hpr_packet_up:packet(),
+    HandlerPid :: pid(),
+    Routes :: [hpr_route:route()]
+) -> ok.
+deliver_packet(_Packet, _HandlerPid, []) ->
+    ok;
+deliver_packet(Packet, HandlerPid, [Route | Routes]) ->
+    case hpr_route:protocol(Route) of
+        router ->
+            hpr_protocol_router:send(Packet, HandlerPid, Route);
+        Protocol ->
+            lager:warning("protocol ~p unsuported", [Protocol])
+    end,
+    deliver_packet(Packet, HandlerPid, Routes).
+
+-spec throttle_check(Packet :: hpr_packet_up:packet()) -> boolean().
+throttle_check(Packet) ->
+    Hotspot = hpr_packet_up:hotspot(Packet),
+    case throttle:check(?HOTSPOT_THROTTLE, Hotspot) of
+        {limit_exceeded, _, _} -> false;
+        _ -> true
+    end.
+
+-spec execute_checks(Packet :: hpr_packet_up:packet(), [{fun(), any()}]) -> ok | {error, any()}.
+execute_checks(_Packet, []) ->
+    ok;
+execute_checks(Packet, [{Fun, ErrorReason} | Rest]) ->
+    case Fun(Packet) of
+        false ->
+            {error, ErrorReason};
+        true ->
+            execute_checks(Packet, Rest)
+    end.
+
+-spec packet_type(Packet :: hpr_packet_up:packet()) ->
+    {join_req, non_neg_integer(), non_neg_integer()} | {uplink, non_neg_integer()} | undefined.
+packet_type(Packet) ->
+    case hpr_packet_up:payload(Packet) of
+        <<?JOIN_REQUEST:3, _:5, AppEUI:64/integer-unsigned-little,
+            DevEUI:64/integer-unsigned-little, _DevNonce:2/binary, _MIC:4/binary>> ->
```

Comment:
@mikev Which library call from `erlang-lorawan` library would you recommend be used here instead?

(We should use destructuring-bind via native Erlang binary as last resort.)